### ISSUE TITLE
Unable to Set-Cookie from within a Strategy ("Cannot set headers after they are sent to the client")

### ISSUE THREAD

**Issue by vakrolme** [nuxt-community/auth-module#474]
### Version

[v4.8.4](https://github.com/auth-module/releases/tag/v4.8.4)

### Reproduction link

[http://none.no](http://none.no)

### Steps to reproduce

I have a cookie-based flow, where Rails API ([jwt_sessions](https://github.com/tuwukee/jwt_sessions#refresh-with-access-token)) sends out an access token in a httpOnly cookie and a CSRF token in the response body. When access token expires, it can be sent to `/refresh` endpoint, which responds with new cookie and tokens. This is made possible by a tweaked local strategy:
```
const DEFAULTS = {
  tokenRequired: true,
  tokenType: false,
  globalToken: true,
  tokenName: 'X-CSRF-TOKEN'
}
```
And a plugin for auth module:
```
$axios.onResponseError(async error => {
      var response = error.response
      if(process.client && $auth.$state.loggedIn && response.status === 401 && response.request._retry != true) {
        const originalRequest = response.config
        originalRequest._retry = true
        const newToken = await $axios.post('/refresh')
        if(newToken.data.csrf) {
          $auth.setUserToken(newToken.data.csrf)
          originalRequest.headers['X-CSRF-TOKEN'] = newToken.data.csrf
          return $axios(originalRequest)
        } else {
          redirect(301, '/login')
        }
      } else {
        return Promise.reject(error)
      }
    });
```
This works in SPA mode, but I also need to recover session on the server when the app is reloaded after access token expires. I want to get old expired tokens from the request cookie, refresh them, fetch the user on server side, and finally pass new tokens to the client. To do this I wrote the following code in strategy:
```
mounted () {
    if (this.options.tokenRequired) {
      const token = this.$auth.syncToken(this.name)
      console.log("synced token is "+token);
      this._setToken(token)

      if (process.server){
        const access = this.$auth.ctx.app.$cookies.get('jwt_access')
        const csrf = this.$auth.ctx.app.$cookies.get('auth._token.jwtSessions')

        if (access && csrf) {
          // setting up retrieved CSRF token
          this.$auth.setToken(this.name, csrf)
          this._setToken(csrf)
          // setting up one-time header with access token
          this.$auth.ctx.app.$axios.setHeader('Authorization', access)

          // fetching user
          this.$auth.ctx.app.$axios.get('/me').then(me_response=>{
            this.$auth.setUser(me_response.data)
          }).catch(err=>{
            // requesting new tokens
            this.$auth.ctx.app.$axios.post('/refresh').then(response=>{
              // setting new tokens to response cookie to be sent to the browser
              this.$auth.ctx.app.$cookies.set('jwt_access', response.data.access, { httpOnly: true, maxAge: 64800, secure: false })
              
              // setting up csrf token & header
              // this.$auth.setToken(this.name, response.data.csrf) // uncommenting this results in "Cannot set headers after they are sent"
              this._setToken(response.data.csrf)

              // setting up one-time access token as a header (trying to rely on cookies resulted in old token being sent)
              this.$auth.ctx.app.$axios.setHeader('Authorization', response.data.access)

              // fetching user with new tokens
              this.$auth.ctx.app.$axios.get('/me').then(new_response => {
                console.log(new_response.data); // returns user when $cookie.set is commented out
                this.$auth.setUser(new_response.data)
              }).catch(err=>{
                console.error(err);
              })
            })
          })
        }
      } else {
        return this.$auth.fetchUserOnce()
      }
    }
  }
```

### What is expected ?

Server response is modified to hold a cookie with a new access token.

### What is actually happening?

When `$cookies.set` is commented out, $axios sends refreshed tokens to the `/me` endpoint and returns the correct user.

But when I uncomment the `$cookies.set` line, it errors:
```
ERROR  Cannot set headers after they are sent to the client

at ServerResponse.setHeader (_http_outgoing.js:485:11)
at p (node_modules/cookie-universal/dist/cookie-universal-common.js:1:1399)
at Object.set (node_modules/cookie-universal/dist/cookie-universal-common.js:1:1836)
at node_modules/cookie-universal/dist/cookie-universal-common.js:1:2103
at Array.forEach (<anonymous>)
at Object.setAll (node_modules/cookie-universal/dist/cookie-universal-common.js:1:1987)
at server.js:955:43
at runMicrotasks (<anonymous>)
at processTicksAndRejections (internal/process/task_queues.js:93:5)
```
Cookie is not set.

Also tried `this.app.ctx.res.setHeader('Set-Cookie', ['jwt_access='+response.data.access+' HttpOnly;']);` — similar result.

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/auth-module/issues/c435">#c435</a>)</em></sub></div>

**Comment by invisor** [comment]
The same problem

**Comment by JohannesLichtenberger** [comment]
Same here

**Comment by JohannesLichtenberger** [comment]
@robsontenorio Aren't you running into the same problem with your Keycloak plugin? At least I'm getting this error, when using your file.

**Comment by robsontenorio** [comment]
@JohannesLichtenberger i am sorry. But I am not running any new project lately. So I am using previous versions of auth-module.

**Comment by vakrolme** [comment]
I solved my issue with this. As I familiarized myself with the internals of the library, it became clear the behavior I saw was not caused by a bug, but rather by the lack of documentation which makes people put logic where it shouldn't go.

The key in my case was to put the refresh logic into axios.js (interceptors) and register it as a general plugin under `plugins` section in the nuxt config, and not as a sub-plugin of `auth` module.  This way it is invoked during auth strategy initialization, so when auth hits `/user` endpoint and receives a 401, axios makes a request to `/refresh`, saves the new token (if the context is server, it is added to an httpOnly response cookie), and then the original query is retied, returning the user object, and the session is restored. The strategy itself is almost identical to what you find in the [default local scheme](https://github.com/nuxt-community/auth-module/blob/dev/lib/schemes/local.js). The only [diff](https://www.diffchecker.com/F6eftbQf) is token name and `_clearToken` method which in my case must clear two tokens instead of one.

I might publish a full repo of my implementation if it would help somebody, but for now I made [a gist](https://gist.github.com/vakrolme/daab84516d17228e8ce89994d33797cf) with the important parts (mind you it's quite bad, and I might have forgot something, but at this point I am just happy it seems to be working). Cheers.
